botcraft/src/AI/Tasks/InventoryTasks.cpp

```diff
@@ -325,19 +325,58 @@ namespace Botcraft
             Position(1, 0, 0), Position(-1, 0, 0)
             });
 
-        // Check if block is air
         bool midair_placing = true;
+        // If no face specified
+        if (!face.has_value())
+        {
+            if (allow_midair_placing) // Then we don't care and default to Up
+            {
+                face = PlayerDiggingFace::Up;
+            }
+            else
+            {
+                std::vector<PlayerDiggingFace> face_candidates; // Faces next to a block
+                for (int face = 0; face < 6; face++)
+                {
+                    const Block *neighbour_block = world->GetBlock(pos + neighbour_offsets[face]);
+                    if (!neighbour_block->GetBlockstate()->IsAir())
+                    {
+                        face_candidates.push_back(static_cast<PlayerDiggingFace>(face));
+                    }
+                }
+                if (face_candidates.size() == 0)
+                {
+                    LOG_WARNING("Can't place a block in midair at " << pos);
+                    return Status::Failure;
+                }
+                Vector3<double> player_orientation;
+                {
+                    std::lock_guard<std::mutex> lock(local_player->GetMutex());
+                    player_orientation = local_player->GetFrontVector();
+                }
+                std::sort( // Find the face face closest to player looking direction
+                    face_candidates.begin(), face_candidates.end(), [&](PlayerDiggingFace a, PlayerDiggingFace b) -> bool
+                    {
+                        Vector3<double> a_offset =  neighbour_offsets[static_cast<int>(a)];
+                        Vector3<double> b_offset = neighbour_offsets[static_cast<int>(b)];
+                        Vector3<double> a_to_player = a_offset - pos; 
+                        Vector3<double> b_to_player = b_offset - pos;
```

Comment:
I think you are missing a 0.5,0.5,0.5 offset here as pos is the int position of the block, i.e. one of the corners, not the center of the cube.